### PR TITLE
Zitadel integration

### PR DESCRIPTION
**Pull request by stankis**

### CODE REVIEW

File: src/types/models/common.ts [hunk 1]
```diff
@@ -13,3 +13,17 @@ export interface BasicRequestParams {
 }
 
 export type ReturningColumns = string | string[];
+
+export type IntrospectionResult = {
+    active: boolean;
+    name?: string;
+};
+
+declare module '@gravity-ui/nodekit' {
```
[the remark this answers is not in the document]
Comment (stankis) [reply src/types/models/common.ts:22]: absolutely agree, moved

File: src/configs/common.ts [hunk 2]
```diff
@@ -16,10 +17,19 @@ export default {
         extended: false,
     },
 
-    appAuthPolicy: AuthPolicy.disabled,
+    appAuthPolicy:
+        Utils.getEnvVariable('ZITADEL') === Zitadel.Enabled
```
[the remark this answers is not in the document]
Comment (stankis) [reply src/configs/common.ts:21]: agree, done

File: src/utils/zitadel.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+export const introspect = async (ctx: AppContext, token?: string): Promise<IntrospectionResult> => {
+    ctx.log('Token introspection');
+
+    try {
+        if (!token) {
+            throw new Error('Token not provided');
+        }
+
+        const response = await axios.post(
```
[the remark this answers is not in the document]
Comment (stankis) [reply src/utils/zitadel.ts:14]: agree, done

File: src/middlewares/auth-zitadel.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import {NextFunction, Request, Response} from '@gravity-ui/expresskit';
```
Comment (imsitnikov) [inline src/middlewares/auth-zitadel.ts:1]: We already have folder for middlwares – `src/components/middlewares`.

File: src/middlewares/auth-zitadel.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        ]);
+
+        ctx.log(`Tokens introspected successfully`);
+
+        if (
+            r1.active &&
+            r2.active &&
+            (r2.name === ZitadelServiceUser.charts || r2.name === ZitadelServiceUser.bi)
+        ) {
+            return next();
```
Comment (imsitnikov) [inline src/middlewares/auth-zitadel.ts:26]: Need to save `r2.name` in `req.locals` and add it into `"info"` context here – `src/components/middlewares/ctx.ts`.

**Review by imsitnikov — CHANGES_REQUESTED** [review]

File: src/middlewares/auth-zitadel.ts (hunk 4 above)
Comment (stankis) [reply src/middlewares/auth-zitadel.ts:1]: agree, moved, didn't find at first glance )

File: src/utils/zitadel.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import {nodekit} from '../nodekit';
+import axiosRetry from 'axios-retry';
+import {Utils} from './utils';
+
+type IntrospectionResult = {
+    active: boolean;
+    name?: string;
+};
+
+axiosRetry(axios, {
```
[the remark this answers is not in the document]
Comment (stankis) [reply src/utils/zitadel.ts:12]: Discussed offline, fixed

File: dev/env/opensource/development.env [hunk 7]
```diff
@@ -5,3 +5,10 @@ MASTER_TOKEN=development-master-token
 CONTROL_MASTER_TOKEN=development-control-master-token
 
 US_SURPRESS_DB_STATUS_LOGS=true
+
+ZITADEL=true
+
+ZITADEL_URI=http://localhost:8085
```
[the remark this answers is not in the document]
Comment (stankis) [reply dev/env/opensource/development.env:11]: of course I will

File: src/middlewares/auth-zitadel.ts (hunk 5 above)
Comment (stankis) [reply src/middlewares/auth-zitadel.ts:26]: Added

**Review by jhoncool — APPROVED** [review]

File: src/const/common.ts [hunk 8]
```diff
@@ -159,6 +159,8 @@ export const AJV_PATTERN_KEYS_NOT_OBJECT = {
 export const US_MASTER_TOKEN_HEADER = 'x-us-master-token';
 export const DL_COMPONENT_HEADER = 'x-dl-component';
 export const DL_WORKBOOK_ID_HEADER = 'x-dl-workbookid';
+export const DL_SERVICE_USER_ACCESS_TOKEN = 'x-dl-service-user-access-token';
```
Comment (jhoncool) [inline src/const/common.ts:162]: Let's add it to appSensitiveKeys, appSensitiveHeaders

**Comment by jhoncool** [comment]
I don't see how can we get some info about current user. For example userId

File: src/const/common.ts (hunk 8 above)
Comment (stankis) [reply src/const/common.ts:162]: done

**Comment by stankis** [comment]
> I don't see how can we get some info about current user. For example userId

r1.name for instance. If we need I may extract more info

**Comment by jhoncool** [comment]
> > I don't see how can we get some info about current user. For example userId
> 
> r1.name for instance. If we need I may extract more info: https://github.com/datalens-tech/datalens-us/pull/116/files#diff-33b5439ed457fcad5bd3cd5f870b8c1fcead6ceded5fe3964d923f5e5f87c718R14

Yes we need userId for DLS

**Comment by stankis** [comment]
> > > I don't see how can we get some info about current user. For example userId
> > 
> > 
> > r1.name for instance. If we need I may extract more info: https://github.com/datalens-tech/datalens-us/pull/116/files#diff-33b5439ed457fcad5bd3cd5f870b8c1fcead6ceded5fe3964d923f5e5f87c718R14
> 
> Yes we need userId for DLS

Added userId to req

**Comment by jhoncool** [comment]
> > > > I don't see how can we get some info about current user. For example userId
> > > 
> > > 
> > > r1.name for instance. If we need I may extract more info: https://github.com/datalens-tech/datalens-us/pull/116/files#diff-33b5439ed457fcad5bd3cd5f870b8c1fcead6ceded5fe3964d923f5e5f87c718R14
> > 
> > 
> > Yes we need userId for DLS
> 
> Added userId to req

In our case it should be in res.locals and received in ctx middleware https://github.com/datalens-tech/datalens-us/blob/main/src/components/middlewares/ctx.ts#L9-L10

**Comment by stankis** [comment]
> > > > > I don't see how can we get some info about current user. For example userId
> > > > 
> > > > 
> > > > r1.name for instance. If we need I may extract more info: https://github.com/datalens-tech/datalens-us/pull/116/files#diff-33b5439ed457fcad5bd3cd5f870b8c1fcead6ceded5fe3964d923f5e5f87c718R14
> > > 
> > > 
> > > Yes we need userId for DLS
> > 
> > 
> > Added userId to req
> 
> In our case it should be in res.locals and received in ctx middleware https://github.com/datalens-tech/datalens-us/blob/main/src/components/middlewares/ctx.ts#L9-L10

I guess this is what you want : https://github.com/datalens-tech/datalens-us/pull/116/files#diff-33b5439ed457fcad5bd3cd5f870b8c1fcead6ceded5fe3964d923f5e5f87c718R26

**Review by jhoncool — APPROVED** [review]

**Review by imsitnikov — APPROVED** [review]

File: dev/env/opensource/development.env [hunk 9]
```diff
@@ -5,3 +5,13 @@ MASTER_TOKEN=development-master-token
...
+
+ZITADEL=false
+
+ZITADEL_URI=http://localhost:8085
+
+CLIENT_ID=269829931184226314@datalens
+CLIENT_SECRET=GRCfRsteLvSqolF122loogYU2oa7LhufWTdS6WCmiOcmSH48rqTmZOKAtLB0A4YT
+
+SERVICE_CLIENT_ID=us
+SERVICE_CLIENT_SECRET=bB6PrbvqgOjnBh5aoTQNZPFE7dysSnSAfZ2Li7vpJwp0AzamSxZNkA0guvkWrtz1
```
Comment (melikhov-dev) [inline dev/env/opensource/development.env:17]: Very common names. It is not clear from the name what they are needed for.

File: src/configs/common.ts [hunk 10]
```diff
@@ -16,10 +17,18 @@ export default {
...
+        ? AuthPolicy.required
+        : AuthPolicy.disabled,
 
     appSensitiveKeys: [US_MASTER_TOKEN_HEADER],
 
+    zitadelEnabled: Utils.isTrueArg(Utils.getEnvVariable('ZITADEL')),
+    zitadelUri: Utils.getEnvVariable('ZITADEL_URI') || 'http://localhost:8080',
+
+    clientId: Utils.getEnvVariable('CLIENT_ID') || '',
+    clientSecret: Utils.getEnvVariable('CLIENT_SECRET') || '',
```
Comment (melikhov-dev) [inline src/configs/common.ts:30]: Very common names. It is not clear from the name what they are needed for.

**Review by jhoncool — APPROVED** [review]

**Review by imsitnikov — APPROVED** [review]

File: dev/env/opensource/development.env (hunk 9 above)
Comment (stankis) [reply dev/env/opensource/development.env:17]: Please suggest new names, will fix within CHARTS-9785

File: src/configs/common.ts (hunk 10 above)
Comment (stankis) [reply src/configs/common.ts:30]: Please suggest new names, will fix within CHARTS-9785

**Pull request merged**